File: hmh-backend/app/models/boq.py
"""BOQ models: BOQHeader, BOQSection, BOQItem."""

import uuid
from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, Enum, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy import DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.base import TimestampMixin
from app.models.enums import BoqStatus, ItemType


class BOQHeader(Base):
    __tablename__ = "boq_headers"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    version_name: Mapped[str] = mapped_column(String(255), nullable=False)
    source_file_name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    source_type: Mapped[str] = mapped_column(String(50), nullable=False, default="manual")
    status: Mapped[BoqStatus] = mapped_column(
        Enum(BoqStatus, name="boq_status_enum", create_type=False),
        nullable=False,
        default=BoqStatus.DRAFT,
    )
    is_active_version: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    uploaded_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    uploaded_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    project: Mapped["Project"] = relationship("Project")  # type: ignore[name-defined]
    sections: Mapped[list["BOQSection"]] = relationship(
        "BOQSection", back_populates="header", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<BOQHeader {self.version_name} (project={self.project_id})>"


class BOQSection(TimestampMixin, Base):
    __tablename__ = "boq_sections"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    boq_header_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("boq_headers.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    stage_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("stage_master.id", ondelete="SET NULL"),
        nullable=True,
    )
    section_name: Mapped[str] = mapped_column(String(255), nullable=False)
    sequence_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    header: Mapped["BOQHeader"] = relationship("BOQHeader", back_populates="sections")
    items: Mapped[list["BOQItem"]] = relationship(
        "BOQItem", back_populates="section", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<BOQSection {self.section_name}>"


class BOQItem(TimestampMixin, Base):
    __tablename__ = "boq_items"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    boq_section_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("boq_sections.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    site_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sites.id", ondelete="SET NULL"),
        nullable=True,
    )
    lot_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("lots.id", ondelete="SET NULL"),
        nullable=True,
    )
    stage_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("stage_master.id", ondelete="SET NULL"),
        nullable=True,
    )
    item_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("items.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    supplier_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("suppliers.id", ondelete="SET NULL"),
        nullable=True,
    )
    raw_description: Mapped[str] = mapped_column(Text, nullable=False)
    normalized_description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    specification: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    item_type: Mapped[ItemType] = mapped_column(
        Enum(ItemType, name="item_type_enum", create_type=False),
        nullable=False,
        default=ItemType.MATERIAL,
    )
    unit: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    planned_quantity: Mapped[Optional[float]] = mapped_column(Numeric(14, 3), nullable=True)
    planned_rate: Mapped[Optional[float]] = mapped_column(Numeric(14, 2), nullable=True)
    # planned_total is GENERATED ALWAYS AS STORED — do NOT write this column
    planned_total: Mapped[Optional[float]] = mapped_column(Numeric(14, 2), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    section: Mapped["BOQSection"] = relationship("BOQSection", back_populates="items")

    def __repr__(self) -> str:
        return f"<BOQItem {self.raw_description[:40]}>"

File: hmh-frontend/src/api/boq.ts
import client from "./client";

export type BoqStatus = "DRAFT" | "UNDER_REVIEW" | "ACTIVE" | "SUPERSEDED" | "ARCHIVED";
export type ItemType = "MATERIAL" | "LABOUR" | "PLANT" | "SERVICE" | "PACKAGE";

export interface BOQHeader {
  id: string;
  project_id: string;
  version_name: string;
  source_file_name: string | null;
  source_type: string;
  status: BoqStatus;
  is_active_version: boolean;
  is_template: boolean;
  template_name: string | null;
  uploaded_by: string | null;
  uploaded_at: string;
  notes: string | null;
}

export interface BOQSection {
  id: string;
  boq_header_id: string;
  stage_id: string | null;
  section_name: string;
  sequence_order: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface BOQItem {
  id: string;
  boq_section_id: string;
  project_id: string;
  site_id: string | null;
  lot_id: string | null;
  stage_id: string | null;
  item_id: string | null;
  supplier_id: string | null;
  raw_description: string;
  normalized_description: string | null;
  specification: string | null;
  item_type: ItemType;
  unit: string | null;
  planned_quantity: number | null;
  planned_rate: number | null;
  planned_total: number | null;
  sort_order: number;
  is_active: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface SectionWithItems {
  section: BOQSection;
  items: BOQItem[];
  section_total: number;
}

export interface FullBOQ {
  header: BOQHeader;
  sections: SectionWithItems[];
  grand_total: number;
}

export interface BOQHeaderCreate {
  version_name: string;
  source_type?: string;
  notes?: string | null;
  is_template?: boolean;
  template_name?: string | null;
}

export interface BOQHeaderUpdate {
  version_name?: string;
  status?: BoqStatus;
  is_active_version?: boolean;
  is_template?: boolean;
  template_name?: string | null;
  notes?: string | null;
}

export interface BOQSectionCreate {
  section_name: string;
  stage_id?: string | null;
  sequence_order?: number;
  notes?: string | null;
}

export interface BOQItemCreate {
  raw_description: string;
  item_type?: ItemType;
  unit?: string | null;
  planned_quantity?: number | null;
  planned_rate?: number | null;
  site_id?: string | null;
  lot_id?: string | null;
  stage_id?: string | null;
  item_id?: string | null;
  supplier_id?: string | null;
  specification?: string | null;
  sort_order?: number;
  notes?: string | null;
}

export interface BOQItemUpdate {
  raw_description?: string;
  item_type?: ItemType;
  unit?: string | null;
  planned_quantity?: number | null;
  planned_rate?: number | null;
  specification?: string | null;
  is_active?: boolean;
  notes?: string | null;
}

export interface LotBOQSummaryItem {
  boq_item_id: string;
  item_id: string | null;
  description: string;
  item_type: ItemType;
  unit: string | null;
  item_name: string;
  allocated_quantity: number;
  used_quantity: number;
  remaining_quantity: number;
  over_quantity: number;
  is_over: boolean;
  planned_rate: number;
  planned_total: number;
  used_cost: number;
}

export interface LotBOQSummary {
  lot_id: string;
  lot_number: string;
  items: LotBOQSummaryItem[];
  total_planned_cost: number;
  total_used_cost: number;
  total_items: number;
  overrun_count: number;
}

// ── Hierarchy types ───────────────────────────────────────────────────────────

export interface TypeBreakdown {
  material_total: number;
  labour_total:   number;
  plant_total:    number;
  service_total:  number;
  other_total?:   number;
}

export interface Variance {
  variance_amount: number;
  variance_pct:    number;
}

export interface MasterSummarySite extends TypeBreakdown, Variance {
  site_id:        string;
  site_name:      string;
  unit_total:     number;
  lot_count:      number;
  site_total:     number;
  item_count:     number;
  has_boq:        boolean;
  boq_header_ids: string[];
}

export interface ProjectMasterSummary extends TypeBreakdown {
  project_id:       string;
  total_planned:    number;
  site_count:       number;
  total_lot_count:  number;
  legacy_lot_count: number;
  sites:            MasterSummarySite[];
}

export interface SiteBOQItem {
  id: string;
  description: string;
  unit: string | null;
  planned_quantity: number;
  planned_rate: number;
  line_total: number;
  item_type: ItemType;
}

export interface SiteBOQSection {
  section_id: string;
  section_name: string;
  sequence_order: number;
  items: SiteBOQItem[];
  section_total: number;
}

export interface SiteBOQSummary extends TypeBreakdown, Variance {
  site_id:           string;
  site_name:         string;
  project_id:        string;
  unit_total:        number;
  lot_count:         number;
  site_total:        number;
  item_count:        number;
  sections:          SiteBOQSection[];
  boq_header_ids:    string[];
  derived_from_lots: boolean;
}

export interface LotBOQRow extends TypeBreakdown, Variance {
  lot_id:         string;
  lot_number:     string;
  lot_status:     string;
  lot_total:      number;
  has_lot_boq:    boolean;
  has_custom_boq: boolean;
  item_count:     number;
  boq_header_ids: string[];
}

// ─────────────────────────────────────────────────────────────────────────────

export interface BOQTemplate {
  id:            string;
  version_name:  string;
  template_name: string | null;
  notes:         string | null;
}

export type BOQApplyMode = "CREATE" | "SAFE" | "FORCE";

export interface BOQTemplatePreviewLot {
  lot_id:                   string;
  lot_number:               string;
  unit_type:                string | null;
  is_freestanding:          boolean;
  is_customized:            boolean;
  has_existing_boq:         boolean;
  existing_item_count:      number;
  new_item_count:           number;
  has_existing_milestones:  boolean;
  existing_milestone_count: number;
  new_milestone_count:      number;
  action:                   "create" | "overwrite" | "skip";
  skip_reason:              string | null;
}

export interface BOQTemplatePreview {
  template_id:            string;
  template_name:          string;
  template_section_count: number;
  template_item_count:    number;
  template_stage_count:   number;
  mode:                   BOQApplyMode;
  lots:                   BOQTemplatePreviewLot[];
  lots_to_apply:          number;
  lots_to_skip:           number;
  lots_needing_overwrite: number;
  total_lots:             number;
}

export interface BOQCloneResult {
  created_count:       number;
  milestones_created:  number;
  deactivated_count:   number;
  skipped_count:       number;
  skipped_reasons:     Record<string, string>;
  lot_ids:             string[];
  freestanding_master: boolean;
  mode:                BOQApplyMode;
}

export const boqApi = {
  listHeaders: async (projectId: string): Promise<BOQHeader[]> => {
    const res = await client.get<{ data: BOQHeader[] }>(`/projects/${projectId}/boq/`);
    return res.data.data;
  },

  createHeader: async (projectId: string, body: BOQHeaderCreate): Promise<BOQHeader> => {
    const res = await client.post<{ data: BOQHeader }>(`/projects/${projectId}/boq/`, body);
    return res.data.data;
  },

  updateHeader: async (projectId: string, headerId: string, body: BOQHeaderUpdate): Promise<BOQHeader> => {
    const res = await client.patch<{ data: BOQHeader }>(`/projects/${projectId}/boq/${headerId}`, body);
    return res.data.data;
  },

  getFullBOQ: async (projectId: string, headerId: string): Promise<FullBOQ> => {
    const res = await client.get<{ data: FullBOQ }>(`/projects/${projectId}/boq/${headerId}/full`);
    return res.data.data;
  },

  markAsTemplate: async (projectId: string, headerId: string, templateName: string): Promise<BOQHeader> => {
    const res = await client.post<{ data: BOQHeader }>(
      `/projects/${projectId}/boq/${headerId}/mark-template`,
      { template_name: templateName },
    );
    return res.data.data;
  },

  seedStandardTemplate: async (projectId: string): Promise<BOQHeader> => {
    const res = await client.post<{ data: BOQHeader }>(
      `/projects/${projectId}/boq/seed-standard-template`,
    );
    return res.data.data;
  },

  listSections: async (headerId: string): Promise<BOQSection[]> => {
    const res = await client.get<{ data: BOQSection[] }>(`/boq/${headerId}/sections/`);
    return res.data.data;
  },

  createSection: async (headerId: string, body: BOQSectionCreate): Promise<BOQSection> => {
    const res = await client.post<{ data: BOQSection }>(`/boq/${headerId}/sections/`, body);
    return res.data.data;
  },

  updateSection: async (headerId: string, sectionId: string, body: { section_name?: string; sequence_order?: number; notes?: string }): Promise<BOQSection> => {
    const res = await client.patch<{ data: BOQSection }>(`/boq/${headerId}/sections/${sectionId}`, body);
    return res.data.data;
  },

  deleteSection: async (headerId: string, sectionId: string): Promise<void> => {
    await client.delete(`/boq/${headerId}/sections/${sectionId}`);
  },

  deleteSectionScoped: async (
    headerId: string,
    sectionId: string,
    scope: "lot" | "site",
  ): Promise<{ scope: string; sections_deleted: number; lots_affected: number; items_deleted: number; message: string }> => {
    const res = await client.delete<{ data: { scope: string; sections_deleted: number; lots_affected: number; items_deleted: number; message: string } }>(
      `/boq/${headerId}/sections/${sectionId}?scope=${scope}`,
    );
    return res.data.data;
  },

  listItems: async (sectionId: string): Promise<BOQItem[]> => {
    const res = await client.get<{ data: BOQItem[] }>(`/boq/sections/${sectionId}/items/`);
    return res.data.data;
  },

  createItem: async (sectionId: string, body: BOQItemCreate): Promise<BOQItem> => {
    const res = await client.post<{ data: BOQItem }>(`/boq/sections/${sectionId}/items/`, body);
    return res.data.data;
  },

  updateItem: async (itemId: string, body: BOQItemUpdate): Promise<BOQItem> => {
    const res = await client.patch<{ data: BOQItem }>(`/boq/items/${itemId}`, body);
    return res.data.data;
  },

  applyItemToAllSiteLots: async (
    itemId: string,
    body: BOQItemUpdate,
  ): Promise<{ updated: number; lots_affected: number; warning: string | null }> => {
    const res = await client.post<{ data: { updated: number; lots_affected: number; warning: string | null } }>(
      `/boq/items/${itemId}/apply-to-all-site-lots`,
      body,
    );
    return res.data.data;
  },

  deleteItem: async (itemId: string): Promise<void> => {
    await client.delete(`/boq/items/${itemId}`);
  },

  importCsv: async (projectId: string, file: File, versionName: string): Promise<BOQHeader> => {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("version_name", versionName);
    const res = await client.post<{ data: BOQHeader }>(
      `/projects/${projectId}/boq/import-csv`,
      formData,
      { headers: { "Content-Type": "multipart/form-data" } },
    );
    return res.data.data;
  },

  getLotBOQSummary: async (lotId: string): Promise<LotBOQSummary> => {
    const res = await client.get<{ data: LotBOQSummary }>(`/lots/${lotId}/boq-summary`);
    return res.data.data;
  },

  getLotHeaders: async (lotId: string): Promise<Array<{
    id: string; project_id: string; version_name: string;
    template_name: string | null; status: string;
    is_active_version: boolean; is_template: boolean;
  }>> => {
    const res = await client.get<{ data: Array<{ id: string; project_id: string; version_name: string; template_name: string | null; status: string; is_active_version: boolean; is_template: boolean }> }>(`/lots/${lotId}/boq-headers`);
    return res.data.data;
  },

  // ── Hierarchy endpoints ──────────────────────────────────────────────────

  masterSummary: async (projectId: string): Promise<ProjectMasterSummary> => {
    const res = await client.get<{ data: ProjectMasterSummary }>(
      `/boq/items/projects/${projectId}/master-summary`,
    );
    return res.data.data;
  },

  siteBoqSummary: async (siteId: string): Promise<SiteBOQSummary> => {
    const res = await client.get<{ data: SiteBOQSummary }>(
      `/boq/items/sites/${siteId}/boq-summary`,
    );
    return res.data.data;
  },

  siteLotBoqs: async (siteId: string): Promise<LotBOQRow[]> => {
    const res = await client.get<{ data: LotBOQRow[] }>(
      `/boq/items/sites/${siteId}/lot-boqs`,
    );
    return res.data.data;
  },

  generateSiteLotBoqs: async (siteId: string): Promise<{
    created: number;
    lot_count: number;
    used_fallback: boolean;
    unassigned_lots: number;
    warning: string | null;
  }> => {
    const res = await client.post<{ data: {
      created: number; lot_count: number;
      used_fallback: boolean; unassigned_lots: number; warning: string | null;
    } }>(
      `/boq/items/sites/${siteId}/generate-lot-boqs`,
    );
    return res.data.data;
  },

  resetLotToSiteBoq: async (lotId: string): Promise<{ created: number }> => {
    const res = await client.post<{ data: { created: number } }>(
      `/boq/items/lots/${lotId}/reset-to-site-boq`,
    );
    return res.data.data;
  },

  // ── Template operations ────────────────────────────────────────────────────

  /** List all global reusable templates (GET /boq-templates/). */
  listTemplates: async (): Promise<BOQTemplate[]> => {
    const res = await client.get<{ data: BOQTemplate[] }>("/boq-templates/");
    return res.data.data;
  },

  /** Dry-run preview — see what will change before applying. */
  previewClone: async (body: {
    template_boq_id: string;
    project_id:      string;
    lot_ids:         string[];
    mode?:           BOQApplyMode;
  }): Promise<BOQTemplatePreview> => {
    const res = await client.post<{ data: BOQTemplatePreview }>(
      "/boq-templates/preview-clone",
      { mode: "CREATE", ...body },
    );
    return res.data.data;
  },

  /** Apply a template to a set of lots. */
  cloneToLots: async (body: {
    template_boq_id:     string;
    project_id:          string;
    lot_ids:             string[];
    mode?:               BOQApplyMode;  // CREATE | SAFE | FORCE
    overwrite?:          boolean;       // deprecated alias for mode="FORCE"
    generate_milestones?: boolean;
  }): Promise<BOQCloneResult> => {
    const res = await client.post<{ data: BOQCloneResult }>(
      "/boq-templates/clone-to-lots",
      { mode: "CREATE", ...body },
    );
    return res.data.data;
  },
};
